nixos/modules/services/continuous-integration/gitlab-runner.nix

```
@@ -1,160 +1,451 @@
 { config, lib, pkgs, ... }:
-
 with lib;
-
 let
   cfg = config.services.gitlab-runner;
-  configFile =
-    if (cfg.configFile == null) then
-      (pkgs.runCommand "config.toml" {
-        buildInputs = [ pkgs.remarshal ];
-        preferLocalBuild = true;
-      } ''
-        remarshal -if json -of toml \
-          < ${pkgs.writeText "config.json" (builtins.toJSON cfg.configOptions)} \
-          > $out
-      '')
-    else
-      cfg.configFile;
   hasDocker = config.virtualisation.docker.enable;
+  hashedServices = with builtins; (mapAttrs' (name: service: nameValuePair
+    "${name}_${config.networking.hostName}_${unsafeDiscardStringContext (
+        substring 0 12
+        (hashString "md5" (toJSON service)))}"
+      service)
+    cfg.services);
+  configPath = "/var/lib/gitlab-runner/.gitlab-runner/config.toml";
+  configureScript = pkgs.writeShellScriptBin "gitlab-runner-configure" (
+    if (cfg.configFile != null) then "cp ${cfg.configFile} ${configPath}" else ''
```

This doesn't work yet.

```gitlab-runner-configure[3294]: cp: cannot create regular file '/var/lib/gitlab-runner/.gitlab-runner/config.toml': No such file or directory```

I think the directory structure is not yet created at this point in time.

```suggestion
    if (cfg.configFile != null) then "mkdir -p /var/lib/gitlab-runner/.gitlab-runner && cp ${cfg.configFile} ${configPath}" else ''
```

Works. Not sure this is the proper way tough.

---

Thanks for a catch! I had `/var/lib/gitlab-runner/.gitlab-runner` created already.